File: kotlin/src/main/kotlin/PlainTextFilesToText.kt
package org.example

import org.example.utils.*
import java.nio.file.Path

fun plainTextFilesToText(uniqueAbsolutePaths: Array<Path>): String? {
  val files = mutableListOf<FTextFilesFileInfo>()

  val handler = fun(file: CompleteFileInfo) {
    files.add(
      FTextFilesFileInfo(
        file = file.file,
        absolutePath = file.absolutePath
      )
    )
  }

  uniqueAbsolutePaths.forEach { path ->
    walkFilterAndHandleFileInfo(path, FileFilterMode.NON_ZERO_BYTE_FILES, FileType.TEXT_FILES, handler)
  }

  if (files.isEmpty()) {
    return null
  }

  val result = StringBuilder()

  return  result.toString()
}

File: go/plain_text_files_to_text.go
package main

import (
	"bufio"
	"os"
	"strings"

	"github.com/markuskreukniet/markus-tools/go/utils"
)

func plainTextFilesToTextToJSON(uniqueFileSystemNodes []utils.FileSystemNode) string {
	text, err := plainTextFilesToText(uniqueFileSystemNodes)
	if err != nil {
		return errorToJSONFunctionResult(err)
	}
	return resultToJSONFunctionResult(text)
}

func plainTextFilesToText(uniqueFileSystemNodes []utils.FileSystemNode) (string, error) {
	var files []utils.FTextFilesFileInfo

	handler := func(file utils.CompleteFileInfo) {
		files = append(files, utils.FTextFilesFileInfo{
			Name:         file.Name,
			AbsolutePath: file.AbsolutePath,
		})
	}

	for _, node := range uniqueFileSystemNodes {
		if err :=
			utils.WalkFilterAndHandleFileInfo(node, utils.NonZeroByteFiles, utils.TextFiles, handler); err != nil {
			return "", err
		}
	}

	if len(files) == 0 {
		return "", nil
	}

	var result strings.Builder

	addNameAndLines := func(fileInfo utils.FTextFilesFileInfo) error {
		if _, err := result.WriteString(fileInfo.Name); err != nil {
			return err
		}

		file, err := os.Open(fileInfo.AbsolutePath)
		if err != nil {
			return err
		}
		defer file.Close()

		scanner := bufio.NewScanner(file)

		for scanner.Scan() {
			if _, err := utils.WriteNewlineString(&result); err != nil {
				return err
			}
			if _, err = result.WriteString(scanner.Text()); err != nil {
				return err
			}
		}

		return scanner.Err()
	}

	if err := addNameAndLines(files[0]); err != nil {
		return "", err
	}

	for _, file := range files[1:] {
		if _, err := utils.WriteTwoNewlineStrings(&result); err != nil {
			return "", err
		}
		if err := addNameAndLines(file); err != nil {
			return "", err
		}
	}

	return result.String(), nil
}
